nat-lab/tests/test_events_link_state.py

```diff
@@ -0,0 +1,489 @@
+
+        for _ in range(0, 5):
+            await asyncio.sleep(1)
+            async with Ping(connection_alpha, beta.ip_addresses[0]).run() as ping:
+                await testing.wait_long(ping.wait_for_next_ping())
+
+        await asyncio.sleep(10)
+        alpha_counts = client_beta.get_link_state_events_count(alpha.public_key)
+        beta_counts = client_alpha.get_link_state_events_count(beta.public_key)
+
+        # There should be no detection in this case
+        # 1 down when node is Connecting and 1 up when node is Connected
+        assert alpha_counts == LinkStateEventsCount(none=0, up=1, down=1)
+        assert beta_counts == LinkStateEventsCount(none=0, up=1, down=1)
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize(
+    "alpha_setup_params",
+    [
+        pytest.param(
+            SetupParameters(
+                connection_tag=ConnectionTag.DOCKER_CONE_CLIENT_1,
+                adapter_type=AdapterType.LinuxNativeWg,
+                connection_tracker_config=generate_connection_tracker_config(
+                    ConnectionTag.DOCKER_CONE_CLIENT_1,
+                    derp_1_limits=ConnectionLimits(1, 1),
+                ),
+                features=TelioFeatures(
+                    no_link_detection=NoLinkDetection(rtt_seconds=1),
+                    wireguard=long_persistent_keepalive_periods(),
+                ),
+            )
+        ),
+    ],
+)
+@pytest.mark.parametrize(
+    "beta_setup_params",
+    [
+        pytest.param(
+            SetupParameters(
+                connection_tag=ConnectionTag.DOCKER_CONE_CLIENT_2,
+                adapter_type=AdapterType.LinuxNativeWg,
+                connection_tracker_config=generate_connection_tracker_config(
+                    ConnectionTag.DOCKER_CONE_CLIENT_2,
+                    derp_1_limits=ConnectionLimits(1, 1),
+                ),
+                features=TelioFeatures(
+                    no_link_detection=NoLinkDetection(rtt_seconds=1),
+                    wireguard=long_persistent_keepalive_periods(),
+                ),
+            )
+        )
+    ],
+)
+async def test_event_link_state_peer_goes_offline(
+    alpha_setup_params: SetupParameters, beta_setup_params: SetupParameters
+) -> None:
+    async with AsyncExitStack() as exit_stack:
+        env = await setup_mesh_nodes(
+            exit_stack, [alpha_setup_params, beta_setup_params]
+        )
+        alpha, beta = env.nodes
+        client_alpha, client_beta = env.clients
+        connection_alpha, connection_beta = [
+            conn.connection for conn in env.connections
+        ]
+
+        async with Ping(connection_alpha, beta.ip_addresses[0]).run() as ping:
+            await testing.wait_long(ping.wait_for_next_ping())
+        async with Ping(connection_beta, alpha.ip_addresses[0]).run() as ping:
+            await testing.wait_long(ping.wait_for_next_ping())
+
+        await client_beta.stop_device()
```

Comment:
Are we sure in this case that the ping was successful before we stop the device? I mean `await` might resolve if the ping was just sent, in that case we might stop the device and enter a race condition where:
- peer's response might have reached us
- peer's device was shut down and the response was not even generated

Overall I wonder why in this testcase the alpha doesn't receive `down` event twice as well? A few lines below it issues a ping and then sleeps, this should guarantee it :thinking:

Reply:
The `wait_for_next_ping` actually waits for the response so it should be all ok.